### ISSUE TITLE
Should Dapr provide an endpoint for structured application logging?

### ISSUE THREAD

**Issue by karolz-ms** [dapr/dapr#3700]
The benefits of structured logging (searchability, avoiding data parsing errors etc) are well-known. Most logging libraries, log shippers such as fluentD, and log storage/indexing systems preserve and use log structure. The weak link is the way how application logs are captured: from the standard output of the service container. This requires the log scraper/shipper to parse the data, making the whole log pipeline less robust and less efficient than necessary.

Dapr could improve on the status quo by exposing an endpoint that accepts logs in well-known, structured format, such as JSON or [MessagePack](https://msgpack.org/). By natively supporting common log destinations via pluggable sinks, Dapr deliver an approachable, complete solution for log shipping that eliminates the data fidelity issues associated with stdout-based logging.

Devil is, as usually, in the details, especially if resiliency and performance are considered. 
- For high performance, logging should use efficient data format(s). Support vary between platforms/logging libraries.
- For high performance, log data should be gathered and formatted in memory and directly sent to log aggregator or final destination in chunks. Cross-process network communication within a pod (how Dapr sidecar works today) is not the best.
- For resiliency, the design should account for situations where log aggregator/sink is not available. Log data should then be cached (with size limits) on the local filesystem (if r/w filesystem is available), and re-played when the log aggregator becomes available.

In other words, Dapr probably should not attempt to become another log shipper like Fluent Bit or Vector ☺️ But at the same time the structured log fidelity is a real issue and current log shippers do not address it well.

@paulyuk FYI

**Comment by daixiang0** [comment]
Maybe not an endpoint, but a flag for structured logs.

**Comment by georgestevens99** [comment]
While I believe it is a waste of resources for Dapr to reinvent logging, I also believe it is an excellent use of resources for Dapr to ENRICH logging, namely through supporting a service (app) to use structured logging that makes dealing with distributed systems so much easier.  If feasible, this will add a lot of value to Dapr.

I encourage doing a feasibility study, starting with Serilog, and conceptually thinking through how this could work for Serilog, and also how it could be extended to other structured logging formats.  Present the feasibility study conceptually.  Get some feedback.  Then decide if it is worth developing further.
